File: projects/python/05-lru-cache/starter/main.py
# 05-lru-cache (starter)
import sys

# TODO(step-1): Node 클래스 정의 (key, val, prev, next)
class Node:
    def __init__(self, key: str = "", val: int = 0):
        self.key = key
        self.val = val
        self.prev = None
        self.next = None

# TODO(step-2): LRUCache 초기화 및 더미 헤드/테일 설정
class LRUCache:
    def __init__(self, capacity: int):
        self.cap = capacity
        self.map = {}
        self.head = Node()
        self.tail = Node()
        self.head.next = self.tail
        self.tail.prev = self.head

    # TODO(step-3): 노드 분리 및 헤드 앞으로 이동
    def _move_to_head(self, node: Node):
        pass

    # TODO(step-4): get 및 put 구현 (축출 로직 포함)
    def get(self, key: str) -> int:
        return -1

    def put(self, key: str, val: int):
        pass

# TODO(step-5): REPL 명령 처리
def main():
    pass

if __name__ == "__main__":
    main()

File: projects/rust/05-lru-cache/starter/main.rs
// 05-lru-cache (starter)
#![allow(unused_variables, dead_code, unused_mut, unused_imports)]
use std::collections::HashMap;
use std::io::{self, BufRead};

// TODO(step-1): CacheNode 인덱스 연결 리스트 구조체
struct CacheNode<K, V> {
    key: K,
    val: V,
    prev: Option<usize>,
    next: Option<usize>,
}

// TODO(step-2): LRUCache 생성자
struct LRUCache<K, V> {
    cap: usize,
    map: HashMap<K, usize>,
    nodes: Vec<CacheNode<K, V>>,
    head: Option<usize>,
    tail: Option<usize>,
}

impl<K: Eq + std::hash::Hash + Clone, V: Clone> LRUCache<K, V> {
    fn new(cap: usize) -> Self {
        LRUCache {
            cap,
            map: HashMap::new(),
            nodes: Vec::new(),
            head: None,
            tail: None,
        }
    }

    // TODO(step-3): 노드 분리 및 헤드 이동
    fn move_to_head(&mut self, idx: usize) {}

    // TODO(step-4): get 및 put 구현
    fn get(&mut self, key: &K) -> Option<V> {
        None
    }

    fn put(&mut self, key: K, val: V) {}
}

// TODO(step-5): REPL CLI
fn main() {
    let stdin = io::stdin();
    for line in stdin.lock().lines() {}
}
